OKTA-658529 dev org minus org2org (#4559)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/secure-oauth-between-orgs/main/index.md

layout: Guides
---

This guide explains how to securely configure Okta hub-and-spoke orgs to synchronize users and groups by using OAuth 2.0 in a [multi-tenant solution](/docs/concepts/multi-tenancy/). The connection between Okta orgs is configured as an OAuth 2.0 [Org2Org app integration](https://help.okta.com/okta_help.htm?type=oie&id=ext-org2org-intg). This connection provides API access to scoped data using the OAuth 2.0 Client Credential flow.

> **Note**: Currently, only the Okta API can be used to enable OAuth 2.0-based provisioning.
---

**Learning outcomes**

> * Set up a hub-and-spoke Okta org provisioning to use the OAuth 2.0 Client Credential grant flow.

Reviewer: ```suggestion
* Set up a hub-and-spoke Okta org provisioning to use the OAuth 2.0 Client Credentials grant flow.
```

* Configure the hub Okta org with service apps for each spoke Okta org.
* Configure an Org2Org app on each spoke Okta org.
* Configure and activate Org2Org provisioning on spoke Okta orgs.
* Rotate keys for the OAuth 2.0 connection.

**What you need**

Multiple Okta orgs for your multi-tenant solution

> **Note:** The Org2Org app integration isn't available in Okta Developer-Edition orgs. If you need to test this feature in your Developer-Edition org, contact your Okta account team.

---

## OAuth 2.0 for secure API connections in multi-tenant solutions

To secure API connections between orgs for a [multi-tenant solution](/docs/concepts/multi-tenancy/) that uses the hub-and-spoke model, use the Okta Org2Org integration as an OAuth 2.0 client.

Previously, the Org2Org integration only supported token-based access to the Okta API.

> The Org2Org integration can now be configured to access Okta APIs as an OAuth 2.0 client using the OAuth 2.0 Client Credential flow. The OAuth 2.0 API access enables you to increase security by limiting the scope of resource access and providing a better mechanism to rotate credentials.

Reviewer: ```suggestion
You can now configure the Org2Org integration to access Okta APIs as an OAuth 2.0 client using the OAuth 2.0 Client Credentials flow. The OAuth 2.0 API access enables you to increase security by limiting the scope of resource access and providing a better mechanism to rotate credentials.
```

<div class="half">

![Displays the service apps required in the hub org and the Org2Org apps required in the spoke org.](/img/multi-tenancy/multi-org-provisioning.png)

</div>

In this configuration:

* The hub org acts as the resource server and the [authorization server](/docs/concepts/auth-servers/)

* Each spoke org represents a [service app in the hub org for the OAuth 2.0 Client Credentials flow](/docs/guides/implement-oauth-for-okta-serviceapp/main/)

* At the spoke org level, an Org2Org app represents the OAuth 2.0 client

> * To limit access to the OAuth 2.0 clients, the hub-org service app needs to be granted with allowed [OAuth 2.0 scopes](https://developer.okta.com/docs/api/oauth2/#okta-admin-management)

Reviewer: ```suggestion
* To limit access to the OAuth 2.0 clients, the hub-org service app needs to be granted with allowed [OAuth 2.0 scopes](https://developer.okta.com/docs/api/oauth2/#okta-admin-management).
```

> After the OAuth 2.0 connection is configured, start to test your connection: push user data from the spoke orgs to the hub org. You can also create a [Rotate key](#key-rotation) schedule to update your credentials regularly.

Reviewer: ```suggestion
After you configure the OAuth 2.0 connection, test your connection: push user data from the spoke orgs to the hub org. You can also create a [Rotate key](#key-rotation) schedule to update your credentials regularly.
```

 }' "https://${yourHubOrgDomain}/oauth2/v1/clients/${yourServiceAppId}"
```

> **Note**: Specify all the required parameters when you update a client app. Partial updates aren't supported. If any mandatory parameters are missing when you update a client app, the update fails. When you update the keys in the service app `jwks` parameter, all the old keys are overwritten. To add a key and keep the old key, you need to specify both old and new keys in the `jwks.keys` array.

### Update the current credentials for the Org2Org app

> From the response of the [previous generate key POST request](#generate-a-new-key-for-the-org2org-app), copy the `kid` property, and [activate the new key by updating the Org2Org app](/docs/reference/api/apps/#update-key-credential-for-application).

Reviewer: ```suggestion
From the response of the [previous generate key POST request](#generate-a-new-key-for-the-org2org-app), copy the `kid` property and [activate the new key by updating the Org2Org app](/docs/reference/api/apps/#update-key-credential-for-application).
```
